### PR TITLE
Add custom session, handle HTTP_CONNECT_ERRORS and retry on ServerDisconnectedError

### PR DESCRIPTION
**Pull request by iMicknl**

### CODE REVIEW

File: custom_components/tahoma/coordinator.py [hunk 1]
```diff
@@ -43,20 +45,26 @@ def __init__(
...
-                if event.name == "DeviceStateChangedEvent":
-                    for state in event.device_states:
-                        device = self.devices[event.deviceurl]
-                        if state.name not in device.states:
-                            device.states[state.name] = state
-                        device.states[state.name].value = self._get_state(state)
+        except (ServerDisconnectedError, NotAuthenticatedException):
+            await self.client.login()
+            self.listener_id = await self.client.register_event_listener()
+            self.devices = await self.client.get_devices(refresh=True)
```
Comment (iMicknl) [inline custom_components/tahoma/coordinator.py:54]: Not sure why the following exception happens.

```
Traceback (most recent call last):
  File "/Users/mick/Projects/personal-projects/home-assistant-core/homeassistant/helpers/update_coordinator.py", line 114, in _handle_refresh_interval
    await self.async_refresh()
  File "/Users/mick/Projects/personal-projects/home-assistant-core/homeassistant/helpers/update_coordinator.py", line 180, in async_refresh
    update_callback()
  File "/Users/mick/Projects/personal-projects/home-assistant-core/homeassistant/helpers/entity.py", line 287, in async_write_ha_state
    self._async_write_ha_state()
  File "/Users/mick/Projects/personal-projects/home-assistant-core/homeassistant/helpers/entity.py", line 308, in _async_write_ha_state
    if not self.available:
  File "/Users/mick/.homeassistant/custom_components/tahoma/tahoma_device.py", line 54, in available
    return self.device.available
  File "/Users/mick/.homeassistant/custom_components/tahoma/tahoma_device.py", line 39, in device
    return self.coordinator.data[self.device_url]
TypeError: list indices must be integers or slices, not str
```

self.devices contains:
```
2020-08-27 00:30:16 WARNING (MainThread) [root] [<pyhoma.models.Device object at 0x11f721510>, <pyhoma.models.Device object at 0x11f721720>, <pyhoma.models.Device object at 0x11f721880>, <pyhoma.models.Device object at 0x11f721930>, <pyhoma.models.Device object at 0x11f7219e0>, <pyhoma.models.Device object at 0x11f721a90>, <pyhoma.models.Device object at 0x11f721b40>, <pyhoma.models.Device object at 0x11f721bf0>, <pyhoma.models.Device object at 0x11f721ca0>, <pyhoma.models.Device object at 0x11f721d50>]
```

File: custom_components/tahoma/__init__.py [hunk 2]
```diff
@@ -76,8 +77,12 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     username = entry.data.get(CONF_USERNAME)
     password = entry.data.get(CONF_PASSWORD)
 
+    session = aiohttp_client.async_create_clientsession(
```
Comment (tetienne) [inline custom_components/tahoma/__init__.py:80]: Can you link where did you see we need this please?

File: custom_components/tahoma/coordinator.py (hunk 1 above)
Comment (tetienne) [inline custom_components/tahoma/coordinator.py:54]: devices here is a dict: deviceurl - > device.

File: custom_components/tahoma/__init__.py (hunk 2 above)
Comment (iMicknl) [reply custom_components/tahoma/__init__.py:80]: https://developers.home-assistant.io/docs/integration_quality_scale_index/#platinum-

File: custom_components/tahoma/coordinator.py (hunk 1 above)
Comment (iMicknl) [reply custom_components/tahoma/coordinator.py:54]: @tetienne what I don't understand is that `self.devices` is a variable that already contains the initial response from `await self.client.get_devices()` in __init__. Thus, this should not be different?

File: custom_components/tahoma/__init__.py (hunk 2 above)
Comment (iMicknl) [reply custom_components/tahoma/__init__.py:80]: Cookie is important to stay logged in and keep the state, afaik.

File: custom_components/tahoma/coordinator.py (hunk 1 above)
Comment (tetienne) [reply custom_components/tahoma/coordinator.py:54]: Nope, have a look here: https://github.com/iMicknl/ha-tahoma/blob/bb28a2bc3d89f7b8ed2c2746a21964450d5fe48f/custom_components/tahoma/coordinator.py#L44

File: custom_components/tahoma/__init__.py (hunk 2 above)
Comment (tetienne) [reply custom_components/tahoma/__init__.py:80]: You are right, I can see a JSESSIONID cookie.

Comment (iMicknl) [reply custom_components/tahoma/__init__.py:80]: @tetienne verify_ssl is optional, I removed this one. However, the CookieJar is required apparently. Without the CookieJar, the cookie is not saved and thus the session is lost.

File: custom_components/tahoma/coordinator.py [hunk 3]
```diff
@@ -43,20 +47,29 @@ def __init__(
...
+            await self.client.login()
+            self.listener_id = await self.client.register_event_listener()
+            self.devices = {
+                d.deviceurl: d for d in await self.client.get_devices(refresh=True)
+            }
             return self.devices
+        except Exception as exception:
+            raise UpdateFailed(exception)
+
+        for event in events:
```
Comment (tetienne) [inline custom_components/tahoma/coordinator.py:64]: events can be None in case we got NotAuthenticatedException.

Comment (iMicknl) [reply custom_components/tahoma/coordinator.py:64]: I do `return self.devices` above, thus this code is never reached in that scenario.

Comment (tetienne) [reply custom_components/tahoma/coordinator.py:64]: Oh indeed sorry, the blank space trick me.

**Review by tetienne — APPROVED** [review]

File: custom_components/tahoma/__init__.py [hunk 4]
```diff
@@ -76,8 +77,12 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
...
+        hass, cookie_jar=CookieJar(unsafe=True)
```
Comment (tetienne) [inline custom_components/tahoma/__init__.py:81]: you can remove cookie_jar parameter. It's useless in our case.

Comment (iMicknl) [reply custom_components/tahoma/__init__.py:81]: I tried, however it doesn't work anymore when I remove this.
https://github.com/iMicknl/ha-tahoma/pull/237#discussion_r478339535

Comment (tetienne) [reply custom_components/tahoma/__init__.py:81]: That's weird. So I'm fine with this PR.

**Pull request merged**